### PR TITLE
Ice Trap

### LINKED ISSUE #56: Ice trap
- [x] Floor tile that covers in ice whatever steps on it/moves over it

### PR DESCRIPTION
**Pull request by kieranmillar**
resolves #56
- New floor tile: Ice Trap, a one-use tile that freezes any entity that steps on it (except Fireball and Iceblock), trapping it inside a new Iceblock
  - Ice Trap added to editor
  - Moving Iceblocks that trigger an Ice Trap just destroy it
  - Protgaonists that are frozen do not record / execute moves while frozen
- New turn event: EntityFrozen
- New SessionAsserter method: assertFloorTileAt()
- Lots of tests covering a range of edge cases

### CODE REVIEW

File: src/GameLogic/Entities/Iceblock.ts [hunk 1]
```diff
@@ -135,6 +135,7 @@ export class Iceblock implements Entity {
 		fireballs.forEach(f => turnState.killEntity(f, TurnEventType.EntityKilled));
 
 		if (level.tilesFloor.get(this.x, this.y) == FloorType.Water) {
+			this.containedEntity = undefined;
```
Comment (EvidentlyCube) [inline src/GameLogic/Entities/Iceblock.ts]: Why this?

File: src/GameLogic/TurnRunner.ts [hunk 2]
```diff
@@ -41,6 +42,31 @@ export class TurnRunner {
 						// @todo When campaignieren are implementieren, implement a way to return that this level has been completed
 					}
 				}
+
+				if (level.tilesFloor.get(i, j) == FloorType.IceTrap) {
```
Comment (EvidentlyCube) [inline src/GameLogic/TurnRunner.ts]: How about a switch statement instead? Though since we plan to refactor it soon anyway let's leave it as is.

File: src/Scenes/GameScene/Views/Editor/Plotter/FloorPlotter.ts [hunk 3]
```diff
@@ -38,6 +38,8 @@ export class FloorPlotter implements Plotter {
 				return textureStore.getTile(GfxConstants.InitialTileset, 6, 7);
 			case FloorType.Exit:
 				return textureStore.getTile(GfxConstants.InitialTileset, 0, 9);
+			case FloorType.IceTrap:
```
Comment (EvidentlyCube) [inline src/Scenes/GameScene/Views/Editor/Plotter/FloorPlotter.ts]: Can you add a `TODO` to #73 to this whole switch statement?

File: src/Scenes/GameScene/Renderers/FloorTilesRenderer.ts [hunk 4]
```diff
@@ -68,6 +68,8 @@ export class FloorTilesRenderer extends PIXI.Sprite {
 				return this._textureStore.getTile(GfxConstants.InitialTileset, 6, 7);
 			case FloorType.Exit:
 				return this._textureStore.getTile(GfxConstants.InitialTileset, 0, 9);
+			case FloorType.IceTrap:
```
Comment (EvidentlyCube) [inline src/Scenes/GameScene/Renderers/FloorTilesRenderer.ts]: Can you add a `TODO` to #73 to this whole switch statement?

File: src/GameLogic/Entities/Iceblock.ts (hunk 1 above)
Comment (kieranmillar) [reply src/GameLogic/Entities/Iceblock.ts]: I figured it might be safer to dereference the contained entity when the iceblock is destroyed, but I guess it doesn't matter. I'll take it out.
